internal/cmd/connector_test.go

```diff
@@ -0,0 +1,269 @@
+package cmd
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"os"
+	"reflect"
+	"sort"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"gotest.tools/v3/assert"
+	"gotest.tools/v3/assert/opt"
+	"gotest.tools/v3/poll"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/infrahq/infra/internal"
+	"github.com/infrahq/infra/internal/certs"
+	"github.com/infrahq/infra/internal/cmd/types"
+	"github.com/infrahq/infra/internal/connector"
+	"github.com/infrahq/infra/internal/server"
+	"github.com/infrahq/infra/internal/server/data"
+	"github.com/infrahq/infra/internal/server/models"
+)
+
+func TestConnector_Run(t *testing.T) {
+	if testing.Short() {
+		t.Skip("too slow for short run")
+	}
+
+	dir := t.TempDir()
+	serverOpts := defaultServerOptions(dir)
+	setupServerOptions(t, &serverOpts)
+	serverOpts.Config = server.Config{
+		Users: []server.User{
+			{Name: "admin@example.com", AccessKey: "0000000001.adminadminadminadmin1234"},
+			{Name: "connector", AccessKey: "0000000002.connectorconnectorconnec"},
+		},
+		Grants: []server.Grant{
+			{User: "user1@example.com", Resource: "testing.ns1", Role: "admin"},
+			{User: "user2@example.com", Resource: "testing", Role: "view"},
+			{Group: "group1@example.com", Resource: "testing.ns1", Role: "logs"},
+		},
+	}
+
+	srv, err := server.New(serverOpts)
+	assert.NilError(t, err)
+
+	fakeKube := &fakeKubeAPI{t: t}
+	kubeSrv := httptest.NewTLSServer(fakeKube)
+	t.Cleanup(kubeSrv.Close)
+
+	ctx := context.Background()
+	runAndWait(ctx, t, srv.Run)
+
+	opts := connector.Options{
+		Server: connector.ServerOptions{
+			URL:                srv.Addrs.HTTPS.String(),
+			AccessKey:          "0000000002.connectorconnectorconnec",
+			TrustedCertificate: serverOpts.TLS.Certificate,
+		},
+		Name:         "testing",
+		CACert:       types.StringOrFile(readFile(t, "testdata/pki/connector.crt")),
+		CAKey:        types.StringOrFile(readFile(t, "testdata/pki/connector.key")),
+		EndpointAddr: types.HostPort{Host: "127.0.0.1", Port: 55555},
+		Kubernetes: connector.KubernetesOptions{
+			AuthToken: "auth-token",
+			Addr:      kubeSrv.URL,
+			CA:        types.StringOrFile(certs.PEMEncodeCertificate(kubeSrv.Certificate().Raw)),
+		},
+		Addr: connector.ListenerOptions{
+			HTTPS:   "127.0.0.1:0",
+			Metrics: "127.0.0.1:0",
+		},
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 20*time.Second)
+	t.Cleanup(cancel)
+
+	runAndWait(ctx, t, func(ctx context.Context) error {
+		return connector.Run(ctx, opts)
+	})
+
+	// check destination has been registered
+	var destination *models.Destination
+	poll.WaitOn(t, func(t poll.LogT) poll.Result {
+		destination, err = data.GetDestination(srv.DB(), data.ByName("testing"))
```

Comment:
Reminder to myself that this will conflict with #3357, and to not try and merge these two PRs at the same time. If that one has merged I need to update this code to use the options struct.